pkg/controller/azvolume.go

```diff
@@ -272,43 +272,39 @@ func (r *ReconcileAzVolume) triggerDelete(ctx context.Context, azVolume *azdiskv
 		var attachments []azdiskv1beta2.AzVolumeAttachment
 		attachments, err = r.cleanUpAzVolumeAttachmentByVolume(deleteCtx, azVolume.Name, azvolume, azureutils.AllRoles, mode)
 		if err == nil {
-			var wg sync.WaitGroup
-			errors := make([]error, len(attachments))
-			numErrors := uint32(0)
+			errorMessageCh := make(chan string)
 
 			// start waiting for replica AzVolumeAttachment CRIs to be deleted
-			for i, attachment := range attachments {
+			for _, attachment := range attachments {
 				var waiter *watcher.ConditionWaiter
 				waiter, err = r.conditionWatcher.NewConditionWaiter(deleteCtx, watcher.AzVolumeAttachmentType, attachment.Name, verifyObjectFailedOrDeleted)
 				if err != nil {
 					break
 				}
 
 				// wait async and report error to go channel
-				wg.Add(1)
-				go func(ctx context.Context, waiter *watcher.ConditionWaiter, i int) {
+				go func(ctx context.Context, waiter *watcher.ConditionWaiter, attachment azdiskv1beta2.AzVolumeAttachment) {
 					defer waiter.Close()
-					defer wg.Done()
 					_, err := waiter.Wait(ctx)
 					if err != nil {
-						errors[i] = err
-						atomic.AddUint32(&numErrors, 1)
+						errorMessageCh <- fmt.Sprintf("%s: %v", attachment.Name, err)
+					} else {
+						errorMessageCh <- ""
 					}
-				}(deleteCtx, waiter, i)
+				}(deleteCtx, waiter, attachment)
 			}
 
-			// start waiting for the attachment clean up to complete (outside of the (if err == nil) to avoid child workflow finishing after parent workflow finishing)
-			wg.Wait()
-
 			if err == nil {
 				// if errors have been found with the wait calls, format the error msg and report via CRI
-				if numErrors > 0 {
-					var errMsgs []string
-					for i, derr := range errors {
-						if derr != nil {
-							errMsgs = append(errMsgs, fmt.Sprintf("%s: %v", attachments[i].Name, derr))
-						}
+				var errMsgs []string
+				for i := 0; i < len(attachments); i++ {
+					v, ok := <-errorMessageCh
+					if ok && v != "" {
+						errMsgs = append(errMsgs, v)
 					}
+				}
+				close(errorMessageCh)
```

Comment:
We need to close the channel unconditionally. With the current structure we can't do this safely because `NewConditionWaiter` could (theoretically) return an error after one or more `Wait` goroutines have already been dispatched. To make this safe, create the `ConditionWaiter` in the goroutine itself. You could then remove the `err == nil` check on line 297 and unconditionally close the channel after iterating through the results.

Reply:
Yes! Done.